xarray/core/computation.py

```diff
@@ -343,8 +344,10 @@ def apply_dataset_ufunc(func, *args, **kwargs):
     list_of_coords = build_output_coords(args, signature, exclude_dims)
 
     args = [getattr(arg, 'data_vars', arg) for arg in args]
+
     result_vars = apply_dict_of_variables_ufunc(
-        func, *args, signature=signature, join=join, fill_value=fill_value)
+        func, *args, signature=signature, join=data_vars_join,
```

Comment:
> 1.) What do you think about removing the join kwarg in the fillna signature in both dataarray.py and dataset.py. The writing of ds0.fillna(ds1) already implies a directionality, so adding a join option can lead to confusion.

Agreed, there isn't really a need for the `join` argument once we have `combine_first`.

For your example `add(ds0, ds2, 'outer', 'outer', dataset_fill_value=np.nan)`, can you clarify what the values of `ds0` and `ds2` are?